components/brave_wallet/renderer/brave_wallet_render_frame_observer.cc

```diff
@@ -0,0 +1,48 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/renderer/brave_wallet_render_frame_observer.h"
+
+#include "content/public/renderer/render_frame.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+
+namespace brave_wallet {
+
+BraveWalletRenderFrameObserver::BraveWalletRenderFrameObserver(
+    content::RenderFrame* render_frame,
+    brave::mojom::DynamicParams dynamic_params)
+    : RenderFrameObserver(render_frame), dynamic_params_(dynamic_params) {
+  native_javascript_handle_.reset(new BraveWalletJSHandler(render_frame));
+}
+
+BraveWalletRenderFrameObserver::~BraveWalletRenderFrameObserver() {}
+
+void BraveWalletRenderFrameObserver::DidStartNavigation(
+    const GURL& url,
+    base::Optional<blink::WebNavigationType> navigation_type) {
+  url_ = url;
+}
+
+void BraveWalletRenderFrameObserver::DidCreateScriptContext(
+    v8::Local<v8::Context> context,
+    int32_t world_id) {
+  if (!dynamic_params_.brave_use_native_wallet || !native_javascript_handle_)
+    return;
+
+  native_javascript_handle_->AddJavaScriptObjectToFrame(context);
+}
+
+void BraveWalletRenderFrameObserver::DidCreateNewDocument() {
+  if (!dynamic_params_.brave_use_native_wallet || !native_javascript_handle_)
+    return;
+
+  native_javascript_handle_->InjectScript();
```

Comment:
if we keep this injected script it needs to go in `DidCreateScriptContext` so it gets created for each isolated world

Reply:
also doesn't this run before `DidCreateScriptContext` which would mean window.ethereum would always be null for the script

Reply:
`DidCreateScriptContext` runs before or on script injection. If you debug it you will see on a callstack that `ExecuteScript` triggers `DidCreateScriptContext` in the same thread if it has not been called before that

Reply:
removed script execution